File: marketdna-backend/app/models/edges.py
"""Pydantic models for the Edge Decay Observatory API."""
from typing import Any, Optional

from pydantic import BaseModel


class EdgeLatest(BaseModel):
    period: str
    edge_ann_pct: Optional[float] = None
    hit_rate: Optional[float] = None
    decile_spread: Optional[float] = None
    n_signals: int
    ci_low: Optional[float] = None
    ci_high: Optional[float] = None


class EdgeSeriesPoint(BaseModel):
    period: str
    edge_ann_pct: Optional[float] = None
    hit_rate: Optional[float] = None
    n_signals: int
    ci_low: Optional[float] = None
    ci_high: Optional[float] = None
    is_backfilled: bool


class EdgeCard(BaseModel):
    edge_key: str
    label: str
    kind: str                              # ranking | event
    blurb: str
    n_readings: int
    status: str                            # HEALTHY | FADING | WEAK | DEAD | TOO_NOISY
    reason: str
    slope: Optional[float] = None          # edge_ann_pct change per reading (trend)
    p_value: Optional[float] = None
    latest: Optional[EdgeLatest] = None
    series: list[EdgeSeriesPoint]


class ObservatoryResponse(BaseModel):
    universe: str
    methodology_version: str
    as_of: str
    edges: list[EdgeCard]
    status_rules: dict[str, Any]


class EdgeMeasurementRow(BaseModel):
    period: str
    window_start: str
    window_end: str
    edge_ann_pct: Optional[float] = None
    hit_rate: Optional[float] = None
    decile_spread: Optional[float] = None
    n_signals: int
    ci_low: Optional[float] = None
    ci_high: Optional[float] = None
    extras: dict[str, Any]
    is_backfilled: bool
    measured_at: str


class FieldHealth(BaseModel):
    edge_key: str
    edge_label: str
    status: str
    reason: str
    latest_edge_ann_pct: Optional[float] = None


class FieldHealthResponse(BaseModel):
    fields: dict[str, FieldHealth]
    as_of: str


class EdgeReportResponse(BaseModel):
    period: str
    universe: str
    methodology_version: str
    as_of: str
    markdown: str


class EdgeHistoryResponse(BaseModel):
    edge_key: str
    label: str
    kind: str
    blurb: str
    universe: str
    methodology_version: str
    status: str
    reason: str
    slope: Optional[float] = None
    p_value: Optional[float] = None
    measurements: list[EdgeMeasurementRow]

File: marketdna-web/src/types/edges.ts
// src/types/edges.ts — Edge Decay Observatory

export type EdgeStatus = 'HEALTHY' | 'FADING' | 'REVIVING' | 'WEAK' | 'DEAD' | 'TOO_NOISY'

export interface FieldHealth {
  edge_key: string
  edge_label: string
  status: EdgeStatus
  reason: string
  latest_edge_ann_pct: number | null
}

export interface FieldHealthResponse {
  fields: Record<string, FieldHealth>
  as_of: string
}

export interface EdgeReportResponse {
  period: string
  universe: string
  methodology_version: string
  as_of: string
  markdown: string
}

export interface EdgeLatest {
  period: string
  edge_ann_pct: number | null
  hit_rate: number | null
  decile_spread: number | null
  n_signals: number
  ci_low: number | null
  ci_high: number | null
}

export interface EdgeSeriesPoint {
  period: string
  edge_ann_pct: number | null
  hit_rate: number | null
  n_signals: number
  ci_low: number | null
  ci_high: number | null
  is_backfilled: boolean
}

export interface EdgeCard {
  edge_key: string
  label: string
  kind: 'ranking' | 'event' | string
  blurb: string
  n_readings: number
  status: EdgeStatus
  reason: string
  slope: number | null
  p_value: number | null
  latest: EdgeLatest | null
  series: EdgeSeriesPoint[]
}

export interface ObservatoryResponse {
  universe: string
  methodology_version: string
  as_of: string
  edges: EdgeCard[]
  status_rules: Record<string, number>
}

export interface EdgeMeasurementRow {
  period: string
  window_start: string
  window_end: string
  edge_ann_pct: number | null
  hit_rate: number | null
  decile_spread: number | null
  n_signals: number
  ci_low: number | null
  ci_high: number | null
  extras: Record<string, unknown>
  is_backfilled: boolean
  measured_at: string
}

export interface EdgeHistoryResponse {
  edge_key: string
  label: string
  kind: string
  blurb: string
  universe: string
  methodology_version: string
  status: EdgeStatus
  reason: string
  slope: number | null
  p_value: number | null
  measurements: EdgeMeasurementRow[]
}
